feat: selector with graph [ARKQB-458]

lucaromagnoli opened the pull request:
## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [x] Update `./docs`
- [x] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [x] Linked issue: https://mckinsey.atlassian.net/browse/ARKQB-458

UI Updates

<img width="765" height="817" alt="Screenshot 2025-11-07 at 16 18 24" src="https://github.com/user-attachments/assets/05380a88-54df-4237-bf3c-35ce3c61d522" />

daniele-marostica committed:
wip: working on mcp over sse implementation

daniele-marostica committed:
wip: fixed client closing connection too early for sse

danielemarostica committed:
fixed connect timeout issue

danielemarostica committed:
Merge branch 'main' into fix/sse-mcp

danielemarostica committed:
wip: fixing tests on ci

danielemarostica committed:
fixed lint

danielemarostica committed:
code review

lucaromagnoli committed:
feat: add graph-constrained selector strategy for teams

- Enhance selector strategy to support optional graph constraints
- Allow multiple edges from same source for selector (unlike strict graph strategy)
- Skip selector agent when only one legal transition exists (optimization)
- Add validateGraphForSelector to allow multiple edges for selector strategy
- Add selectMemberWithFilter to constrain selector to legal transitions
- Add comprehensive tests for validation and selector logic

This enables teams to use graph-defined workflow constraints while
maintaining AI-driven decision making through the selector agent.

lucaromagnoli committed:
Merge remote-tracking branch 'origin/main' into feat/graph-selector-hybrid

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
fix: initialize ResourceValidator in team webhook tests

Initialize ResourceValidator with fake client in team webhook tests to prevent nil pointer panic when validating selector agent.

lucaromagnoli committed:
fix: create all referenced agents in team webhook test setup

Create all agents referenced as team members (researcher, analyst, writer) in addition to the coordinator agent to prevent validation failures.

lucaromagnoli committed:
fix: address linting issues and improve function naming

- Fix ginkgo-linter errors: use HaveOccurred()/ToNot(HaveOccurred()) instead of BeNil()
- Fix goconst: extract string literals as constants (ToolTypeAgent, StrategySelector)
- Fix unparam: remove unused participantsList parameter from selectMember
- Fix gocritic: refactor if-else chains to switch statements
- Reduce cognitive complexity: extract selectNextMember helper functions
- Rename selectMemberWithFilter to selectMemberWithConstraints for clarity
- Add nolint comment for remaining acceptable complexity

lucaromagnoli committed:
refactor: reorder functions to place selectMemberWithConstraints after selectMember

Improve code organization by placing selectMemberWithConstraints directly after selectMember, since selectMember is a wrapper that calls it.

lucaromagnoli committed:
refactor: make selectMember accept candidateMembers parameter

Refactor selectMember to accept candidateMembers and candidateIndices parameters:
- nil = select from all members (no constraints)
- provided = select from constrained list (with constraints)

This makes the API explicit - constraints are applied via parameters rather than separate functions.

lucaromagnoli committed:
refactor: restore selectMember signature close to main with optional constraint parameters

Keep selectMember signature similar to main (accepts participantsList/rolesList strings), but add optional candidateMembers/candidateIndices parameters for constraints. This minimizes the diff while adding constraint support.

lucaromagnoli committed:
chore: remove docs review files

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
test: add unit tests and Chainsaw e2e test for graph+selector strategy

- Add unit tests for helper functions (buildHistory, buildParticipants, buildRoles) - 100% coverage
- Add unit tests for selectNextMember and selectNextMemberWithGraphConstraints
- Create Chainsaw e2e test for graph-constrained selector strategy
- Coverage improved from 11.2% to 12.6%

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
fix: resolve linting errors

- Replace string literal 'agent' with ToolTypeAgent constant in context_utils.go
- Remove unused mockSelectorAgent type and methods from team_selector_test.go
- Fix gofmt formatting issues

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

havasik commented on ark/internal/genai/agent.go:
```diff
@@ -266,7 +266,7 @@ func (a *Agent) GetName() string {
 
 func (a *Agent) GetType() string {
-	return "agent"
+	return ToolTypeAgent
```
would be great to standardise these strings but ToolTypeAgent is I believe is specifically for Tool types

havasik commented on ark/internal/genai/context_utils.go:
```diff
@@ -56,7 +56,7 @@ func WithExecutionMetadata(ctx context.Context, metadata map[string]interface{})
 		case "team":
 			tmpCtx = context.WithValue(tmpCtx, teamKey, value)
-		case "agent":
+		case ToolTypeAgent:
```
same

havasik commented on ark/internal/genai/team_selector.go:
```diff
@@ -159,15 +243,30 @@ func (t *Team) executeSelector(ctx context.Context, userInput Message, history [
+
+	// Build member map for quick lookup
+	memberMap := make(map[string]TeamMember)
+	memberIndexMap := make(map[string]int)
```
I don't think we need this index map and all the complexity it brings, the last parameter of t.executeMemberAndAccumulate might be obsolete

dwmkerr commented:
This looks great, we'd need to update the docs in the team content somewhere to discuss how this works, I wonder if there is value in a diagram for this one or the selector

havasik commented on ark/internal/genai/team_selector.go:
```diff
@@ -123,28 +144,91 @@ func (t *Team) selectMember(ctx context.Context, messages []Message, tmpl *templ
+	}
 }
 
+//nolint:gocognit // Complex function orchestrating selector logic with graph constraints, but cohesive responsibilities
```
Much of the complexity could be reduced by consistently using TeamMember objects instead of string, the whole overhead of passing over memberMap, memberIndexMap etc.

havasik commented on ark/internal/genai/team_selector.go:
```diff
@@ -159,15 +243,30 @@ func (t *Team) executeSelector(ctx context.Context, userInput Message, history [
 
-		nextMember, memberIndex, err := t.selectMember(ctx, messages, tmpl, participantsList, rolesList, previousMember)
+		// Determine next member based on graph constraints (if any)
+		nextMember, memberIndex, err := t.selectNextMember(ctx, messages, tmpl, previousMember, legalTransitions, memberMap, memberIndexMap)
```
Could be bit hierarchically confusing that this function is called `selectNextMember` and then there's a `selectMember` function on the same logical abstraction level as `selectNextMemberWithGraphConstraints`

lucaromagnoli committed:
refactor: address PR review comments for graph-constrained selector

- Create MemberTypeAgent constant separate from ToolTypeAgent
- Remove memberIndexMap complexity, use TeamMember objects directly
- Simplify function signatures to work with TeamMember objects
- Fix function naming hierarchy:
  - selectNextMember -> determineNextMember (orchestrator)
  - selectNextMemberWithGraphConstraints -> selectMemberWithGraphConstraints
  - selectMember -> selectMemberWithSelectorAgent
- Update legalTransitions to use map[string][]TeamMember instead of strings
- Add findMemberIndex helper for telemetry index calculation
- Update all tests to match new function signatures and data structures

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
docs: add documentation for graph-constrained selector strategy

- Add new section in team-strategies.mdx explaining graph-constrained selector
- Update mermaid diagram to include graph-constrained selector visualization
- Add example workflows showing how graph constraints work with selector
- Update reference docs to mention selector + graph combination
- Add section in team-selector-strategy.mdx with example and link

Addresses PR review comment about needing documentation updates.

lucaromagnoli committed:
refactor: restore original function names while keeping improvements

- Restore selectNextMember (was determineNextMember)
- Restore selectNextMemberWithGraphConstraints (was selectMemberWithGraphConstraints)
- Restore selectMember (was selectMemberWithSelectorAgent) with updated signature
- Keep improvements: removed memberIndexMap, using TeamMember objects, map[string][]TeamMember
- Update selectMember signature to accept participantsList, rolesList, candidateMembers

lucaromagnoli committed:
fix: restore original return type (TeamMember, int, error)

Restore the original return type signature for selectNextMember and
selectNextMemberWithGraphConstraints to return (TeamMember, int, error)
instead of (TeamMember, error). The index is calculated using findMemberIndex
when needed but returned as part of the function signature to match original.

lucaromagnoli committed:
refactor: reduce cognitive complexity in selectMember function

Extract selectMember logic into three helper functions:
- prepareMemberSearchData: prepares member lists and indices
- executeSelectorAgent: handles selector agent execution and parsing
- findSelectedMember: finds member with fallback logic

This reduces cognitive complexity from 21 to acceptable levels while
maintaining the original function signatures and behavior.

lucaromagnoli committed:
refactor: simplify selectMember to accept optional candidateMembers parameter

- Remove duplicate selectMemberWithConstraints function
- Add optional candidateMembers parameter to selectMember
- Use candidateMembers when provided, otherwise use all team members
- Reduces code duplication and cognitive load

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
refactor: rename functions for clarity

- Rename selectNextMember to determineNextMember (router function)
- Rename selectNextMemberWithGraphConstraints to selectFromGraphConstraints (selection logic)
- Update function documentation to reflect their actual roles
- Update test function names to match

lucaromagnoli committed:
perf: use lookup maps to avoid repeated iterations for member indices

- Build memberIndexMap once in selectMember() for O(1) index lookups
- Build memberLookup and memberIndexMap once in selectFromGraphConstraints()
- Eliminates O(n) iterations over t.Members for each index lookup
- Improves performance from O(n) per lookup to O(1) after initial O(n) map build

lucaromagnoli committed:
perf: add GetIndex() to TeamMember interface to eliminate map lookups

- Add GetIndex() method to TeamMember interface
- Create indexedTeamMember wrapper that stores index when loading team members
- Update Agent and Team to implement GetIndex() (returns -1 when not wrapped)
- Replace all memberIndexMap lookups with direct member.GetIndex() calls
- Eliminates O(n) map building and lookups in favor of O(1) direct access
- Update all test mocks to include index field

lucaromagnoli committed:
refactor: remove GetIndex from TeamMember interface and simplify return types

- Remove GetIndex() method from TeamMember interface
- Remove indexedTeamMember wrapper
- Change selection functions to return (TeamMember, error) instead of (TeamMember, int, error)
- Update tests to remove index-related assertions

lucaromagnoli committed:
feat: expose graph-constrained selector strategy in UI and CLI

- Allow graph edges to be configured when using selector strategy in team editor
- Add validation for optional graph edges with selector strategy
- Update CLI team generator to mention graph constraints option
- Remove review documentation files from docs/reviews/

lucaromagnoli committed:
docs: document graph-constrained selector in API and add test

- Update create_team docstring to mention selector can be combined with graph constraints
- Add test case for graph+selector combination to verify API support

lucaromagnoli committed:
feat: clarify selector strategy supports graph constraints in UI

- Update strategy dropdown to show 'Selector (can add graph constraints)'
- Add informational box explaining graph constraints can be added to selector strategy

lucaromagnoli committed:
fix: correct graph edge field name in test assertions

- Change from_ to from in test assertions to match JSON alias
- GraphEdge model uses from_ as Python field with 'from' as JSON alias
- API responses use 'from' in JSON, not 'from_'

lucaromagnoli committed:
fix: prettier formatting in team-editor.tsx

- Reorder CSS classes to match Prettier rules
- Fix text line breaks for proper formatting

lucaromagnoli commented:
@all-contributors please add @lucaromagnoli for code,test,doc

lucaromagnoli committed:
chore: update generated types to match API documentation

- Add execution strategies documentation to Create Team endpoint
- Generated from updated API docstring

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid

lucaromagnoli committed:
Merge branch 'main' into feat/graph-selector-hybrid
